### ISSUE TITLE
API to get running Duktape version

### ISSUE THREAD

**Issue by fatcerberus** [svaarala/duktape#607]
Something like `duk_get_version()` or `duk_get_git_describe()` which returned the `git describe` string for the running Duktape version.  This could be useful when dynamic linking against Duktape as a way to get the actual version of the library (instead of just what `duktape.h` thinks it is).

Most libraries I've encountered include a function like this, even e.g. zlib, Duktape seems to be one of the few outliers.

**Comment by svaarala** [comment]
I guess this would just return `DUK_VERSION`?

**Comment by fatcerberus** [comment]
I've actually find `DUK_GIT_DESCRIBE` to be more useful for logging and identification (since that differentiates prereleases).  Usually APIs of this kind return strings.

**Comment by svaarala** [comment]
But comparing `DUK_GIT_DESCRIBE` to check for compatibility would be quite difficult?

**Comment by fatcerberus** [comment]
This would also be useful when using Duktape from other languages than C which can't include the duktape.h. :)

**Comment by svaarala** [comment]
There's also a typical difference between libraries intended to be external vs. libraries typically embedded in a program. Duktape is more typically embedded than used as a DLL, though both options are possible.

**Comment by fatcerberus** [comment]
That's true, I guess a function that returned DUK_VERSION would be sufficient for verifying compatibility, although I'd personally really like the `git describe` string to be available somehow too.

**Comment by svaarala** [comment]
This is of course easy to implement yourself if you're linking against a DLL you compile. A third party DLL is another issue of course.

**Comment by fatcerberus** [comment]
The main use case I'm thinking of is, e.g. Linux packaging where everything is dynamically linked so that you don't know what version of anything is installed unless you actually query the library.  And maybe the person using Duktape is different from the person who packaged it... :)

**Comment by svaarala** [comment]
But in that case the system provides both `duk_config.h`, `duktape.h` and the DLL.

If it doesn't, anything could happen.

**Comment by svaarala** [comment]
For example, if `duk_config.h` doesn't match the DLL the library and the application (if compared against a different configuration) can be binary incompatible. Same goes for `duktape.h`.

**Comment by fatcerberus** [comment]
That's true, although devil's advocate: what if the caller isn't a C/C++ program?  For example calling into Duktape using .NET P/Invoke.

I'm mostly indifferent to this, I just think it would be nice for consistency with other libraries.  If you don't want to implement it for footprint reasons that's fine too. :)

**Comment by svaarala** [comment]
Those environments always need to deal with some API calls being actual macros etc, so to remain versionable they must "wrap" everything anyway, so why not `DUK_VERSION`?

**Comment by svaarala** [comment]
It's not for footprint reasons, I'm just trying to figure out what actual problem is being solved.

**Comment by fatcerberus** [comment]
The main concrete use case I'm thinking of here is one where someone compiles Duktape as a DLL for use alongside another program, but the Duktape functionality is optional so that the program loads it dynamically (using dlopen, LoadLibrary, what-have-you).

Now maybe the person writing that app wants to allow a new version of the DLL to be dropped in later--to get security fixes, say.  The only way that program can know the *real* Duktape version then is, as opposed to what its bindings say it is, is to query the library itself.

**Comment by svaarala** [comment]
Ok, so basically this could only be a patch level update?

**Comment by fatcerberus** [comment]
Well, since Duktape uses semver it *could* theoretically be a minor increment too. :) In practice probably not though given the macro <-> function flexibility.

**Comment by svaarala** [comment]
Minor updates may change file/macro separation arbitrarily so they're compile, but not binary, compatible.
